crates/electrum/src/electrum_ext.rs

```diff
@@ -245,54 +127,129 @@ impl<A: ElectrumApi> ElectrumExt for A {
                 })
                 .collect::<BTreeMap<_, _>>();
 
-            break (
-                ElectrumUpdate {
-                    chain_update,
-                    relevant_txids,
-                },
-                keychain_update,
-            );
+            break FullScanResult {
+                graph_update,
+                chain_update,
+                last_active_indices: keychain_update,
+            };
         };
 
-        Ok((electrum_update, keychain_update))
+        Ok(update)
     }
 
     fn sync(
         &self,
-        prev_tip: CheckPoint,
-        misc_spks: impl IntoIterator<Item = ScriptBuf>,
-        txids: impl IntoIterator<Item = Txid>,
-        outpoints: impl IntoIterator<Item = OutPoint>,
+        request: SyncRequest,
         batch_size: usize,
-    ) -> Result<ElectrumUpdate, Error> {
-        let spk_iter = misc_spks
-            .into_iter()
-            .enumerate()
-            .map(|(i, spk)| (i as u32, spk));
-
-        let (mut electrum_update, _) = self.full_scan(
-            prev_tip.clone(),
-            [((), spk_iter)].into(),
-            usize::MAX,
-            batch_size,
-        )?;
+    ) -> Result<SyncResult<ConfirmationHeightAnchor>, Error> {
+        let mut tx_cache = request.tx_cache.clone();
 
-        let (tip, _) = construct_update_tip(self, prev_tip)?;
+        let full_scan_req = FullScanRequest::from_chain_tip(request.chain_tip.clone())
+            .cache_txs(request.tx_cache)
+            .set_spks_for_keychain((), request.spks.enumerate().map(|(i, spk)| (i as u32, spk)));
+        let mut full_scan_res = self.full_scan(full_scan_req, usize::MAX, batch_size)?;
+
+        let (tip, _) = construct_update_tip(self, request.chain_tip)?;
         let cps = tip
             .iter()
             .take(10)
             .map(|cp| (cp.height(), cp))
             .collect::<BTreeMap<u32, CheckPoint>>();
 
-        populate_with_txids(self, &cps, &mut electrum_update.relevant_txids, txids)?;
+        populate_with_txids(
+            self,
+            &cps,
+            &mut tx_cache,
+            &mut full_scan_res.graph_update,
+            request.txids,
+        )?;
+        populate_with_outpoints(
+            self,
+            &cps,
+            &mut tx_cache,
+            &mut full_scan_res.graph_update,
+            request.outpoints,
+        )?;
+
+        Ok(SyncResult {
+            chain_update: full_scan_res.chain_update,
+            graph_update: full_scan_res.graph_update,
+        })
+    }
+}
+
+/// Trait that extends [`SyncResult`] and [`FullScanResult`] functionality.
+///
+/// Currently, only a single method exists that converts the update [`TxGraph`] to have an anchor
+/// type of [`ConfirmationTimeHeightAnchor`].
+pub trait ElectrumResultExt {
```

Comment:
@LagginTimes do you have any ideas on this one?

Reply:
The reason why `sync`/`full_scan` returns updates with `ConfirmationHeightAnchor` is because this is what the Electrum API provides us with. To get the block time requires extra calls, hence, I think it makes sense to put this in a separate method. I agree that using the `ElectrumResultExt` trait is clunky. Maybe a better option will be to first return a custom type? And methods on the custom type returns `SyncResult`/`FullScanResult` per the caller's requirement.

Reply:
@ValuedMammal I think you will be happy with this solution: https://github.com/bitcoindevkit/bdk/pull/1403/commits/7a69d359135efc2898c318e90479dd03791962ea

Reply:
Maybe I'm bikeshedding, but what if we just made this another `ElectrumExt` method that does the same thing as the current `try_into_confirmation_time_result` and merely converts the first graph update to a new TxGraph containing confirmation time anchors, with no need to introduce a new type.

```rust
    /// Converts the given [`TxGraph`] into one with anchors of type [`ConfirmationTimeHeightAnchor`].
    fn confirmation_time_tx_graph(
        &self,
        graph: TxGraph<ConfirmationHeightAnchor>,
    ) -> Result<TxGraph<ConfirmationTimeHeightAnchor>, Error>;
```